reactioncommerce/reaction
# refactor: add isSoldOut prop to variants and options in Catalog

Impact: minor
Type:  refactor

## Summary
Sometimes in is necessary and convenient to have inventory information at the variant and option level.

Use case 1: In product listing page, it should be possible not to render or disable sold out variants.

Use case 2. In product detail page it should be possible to not render or disable sold out variants or options.

Use case 3: When indexing data, some systems such as Elasticsearch monitor the `Catalog` collection to update the index. If inventory information is not present in the `Catalog` collection, it will not be synced to the search index.

## Testing
Test 1
1. start Reaction and verify migration 68 completed successfully.
2. Verify the `isSoldOut` props is present in the variants and options in the `Catalog` collection.
3. Verify value of `isSoldOut` prop is correct. If setting `canSellVariantWithoutInventory` is set to true, which it is by default, then `isSoldOut` should be `false`. If the setting is set to false, and inventory tracking is enabled for a variant or option, then it should be reflective of the quantity value. If `quantity > 0`, then `isSoldOut` = false, if `quantity == 0`, then `isSoldOut` should be true for that variant or option.

Test 2
1. Enable inventory tracking for an option
2. Set quantity to 0
3. Verify product option is sold out in the storefront

Test 3
1. Enable inventory tracking for a product with at least on variant and two options.
2. Set inventory to 0 for all options of a variant
3. Verify variants and options are sold out, (disabled) in storefront.
